File: packages/core/src/patch.ts
// First-class patch artifact for the base + patch layered architecture.
//
// A Patch is the human-editable, git-committable form of a WorkingTree diff.
// It differs from the internal TreeDiff in two ways:
//
//   1. Addresses nodes by SmokePing path ("/CDN/Cloudflare") rather than by
//      internal id ("c:CDN/Cloudflare"), so git diffs stay readable and two
//      users can PR each other's patches.
//   2. Pins a `baseVersion` so the CLI / web can detect drift when the
//      upstream catalogue evolves.
//
// This module also owns the v:2 URL-hash format (compressed patch YAML) and
// provides unified browser helpers that try v:2 first and fall back to the
// legacy v:1 TreeDiff JSON so old share links keep working.

import { compressToEncodedURIComponent, decompressFromEncodedURIComponent } from 'lz-string';
import { parse as yamlParse, stringify as yamlStringify } from 'yaml';

import type {
  Catalog,
  CatalogVersion,
  Language,
  Node,
  NodeType,
  Probe,
  RootMeta,
  WorkingTree,
} from './types.js';
import { idToPath, pathToId } from './tree.js';
import {
  MAX_URL_LENGTH,
  URL_HASH_PREFIX,
  applyDiff,
  computeDiff,
  decodeTree,
  type CustomEntry,
  type NodeOverride,
  type TreeDiff,
} from './url-state.js';

// ---------------------------------------------------------------------------
// Schema
// ---------------------------------------------------------------------------

export const PATCH_SCHEMA = 1;
export const PATCH_URL_V = 2;

// Shape of a user-added node as it appears in a patch. Omits internal fields
// (id, source) — applyPatch regenerates those on hydration.
export interface PatchNode {
  type: NodeType;
  name: string;
  menu: string;
  title: string;
  included?: boolean;
  host?: string;
  probe?: Probe;
  comparisonChildren?: string[];
  extra?: Record<string, string>;
  children?: PatchNode[];
}

export interface PatchCustomEntry {
  parentPath: string | null;
  node: PatchNode;
}

export interface Patch {
  schema: typeof PATCH_SCHEMA;
  baseVersion: CatalogVersion;
  language?: Language;
  root?: Partial<RootMeta>;
  excluded?: string[];
  overrides?: Record<string, NodeOverride>;
  custom?: PatchCustomEntry[];
}

export interface DriftReport {
  missingPaths: string[];
  baseMismatch: { patch: CatalogVersion; actual?: CatalogVersion } | null;
}

// ---------------------------------------------------------------------------
// encode / apply
// ---------------------------------------------------------------------------

export function encodePatch(tree: WorkingTree, base: Catalog): Patch {
  const diff = computeDiff(tree, base);
  const patch: Patch = {
    schema: PATCH_SCHEMA,
    baseVersion: base.version ?? { date: 'unknown', sha: 'unknown' },
  };

  if (diff.lang) patch.language = diff.lang;
  if (diff.root) patch.root = diff.root;

  if (diff.ex) {
    const paths = diff.ex
      .map((id) => idToPath(base.nodes, id))
      .filter((p): p is string => p !== null);
    if (paths.length) patch.excluded = paths;
  }

  if (diff.ov) {
    const ov: Record<string, NodeOverride> = {};
    for (const [id, override] of Object.entries(diff.ov)) {
      const path = idToPath(base.nodes, id);
      if (path) ov[path] = override;
    }
    if (Object.keys(ov).length) patch.overrides = ov;
  }

  if (diff.cu && diff.cu.length) {
    patch.custom = diff.cu.map((entry) => ({
      parentPath: entry.parentId ? idToPath(tree.nodes, entry.parentId) : null,
      node: toPatchNode(entry.node),
    }));
  }

  return patch;
}

export function applyPatch(patch: Patch, base: Catalog): { tree: WorkingTree; drift: DriftReport } {
  const drift: DriftReport = { missingPaths: [], baseMismatch: null };

  if (base.version && patch.baseVersion && base.version.sha !== patch.baseVersion.sha) {
    drift.baseMismatch = { patch: patch.baseVersion, actual: base.version };
  }

  const diff: TreeDiff = { v: 1 };
  if (patch.language) diff.lang = patch.language;
  if (patch.root) diff.root = patch.root;

  if (patch.excluded) {
    const ids: string[] = [];
    for (const p of patch.excluded) {
      const id = pathToId(base.nodes, p);
      if (id) ids.push(id);
      else drift.missingPaths.push(p);
    }
    if (ids.length) diff.ex = ids;
  }

  if (patch.overrides) {
    const ov: Record<string, NodeOverride> = {};
    for (const [path, override] of Object.entries(patch.overrides)) {
      const id = pathToId(base.nodes, path);
      if (id) ov[id] = override;
      else drift.missingPaths.push(path);
    }
    if (Object.keys(ov).length) diff.ov = ov;
  }

  if (patch.custom) {
    const cu: CustomEntry[] = [];
    for (const entry of patch.custom) {
      let parentId: string | null = null;
      if (entry.parentPath) {
        const resolved = pathToId(base.nodes, entry.parentPath);
        if (resolved === null) {
          drift.missingPaths.push(entry.parentPath);
          // Attach to root as a best-effort fallback.
        } else {
          parentId = resolved;
        }
      }
      cu.push({ parentId, node: fromPatchNode(entry.node) });
    }
    if (cu.length) diff.cu = cu;
  }

  const tree = applyDiff(diff, base);
  return { tree, drift };
}

function toPatchNode(node: Node): PatchNode {
  const p: PatchNode = {
    type: node.type,
    name: node.name,
    menu: node.menu,
    title: node.title,
  };
  if (!node.included) p.included = false;
  if (node.host !== undefined) p.host = node.host;
  if (node.probe) p.probe = node.probe;
  if (node.comparisonChildren) p.comparisonChildren = node.comparisonChildren;
  if (node.extraAttrs && Object.keys(node.extraAttrs).length) p.extra = node.extraAttrs;
  if (node.children.length) p.children = node.children.map(toPatchNode);
  return p;
}

function fromPatchNode(p: PatchNode): Node {
  const n: Node = {
    id: `x:${crypto.randomUUID()}`,
    source: 'custom',
    type: p.type,
    name: p.name,
    menu: p.menu,
    title: p.title,
    included: p.included ?? true,
    children: p.children ? p.children.map(fromPatchNode) : [],
  };
  if (p.host !== undefined) n.host = p.host;
  if (p.probe) n.probe = p.probe;
  if (p.comparisonChildren) n.comparisonChildren = p.comparisonChildren;
  if (p.extra) n.extraAttrs = p.extra;
  return n;
}

// ---------------------------------------------------------------------------
// YAML I/O
// ---------------------------------------------------------------------------

export function patchToYaml(patch: Patch): string {
  return yamlStringify(patch, { lineWidth: 100 });
}

export function patchFromYaml(text: string): Patch {
  const parsed = yamlParse(text);
  if (!parsed || typeof parsed !== 'object') {
    throw new Error('patchFromYaml: input is not a YAML mapping');
  }
  const p = parsed as Partial<Patch>;
  if (p.schema !== PATCH_SCHEMA) {
    throw new Error(
      `patchFromYaml: unsupported schema (got ${String(p.schema)}, expected ${PATCH_SCHEMA})`,
    );
  }
  if (!p.baseVersion || typeof p.baseVersion.sha !== 'string') {
    throw new Error('patchFromYaml: missing or malformed baseVersion');
  }
  return parsed as Patch;
}

// ---------------------------------------------------------------------------
// URL hash — v:2 carries compressed patch YAML, v:1 (legacy TreeDiff JSON)
// still decoded for back-compat.
// ---------------------------------------------------------------------------

interface HashEnvelopeV2 {
  v: typeof PATCH_URL_V;
  y: string;
}

export function encodePatchToHash(patch: Patch): string {
  const envelope: HashEnvelopeV2 = { v: PATCH_URL_V, y: patchToYaml(patch) };
  return compressToEncodedURIComponent(JSON.stringify(envelope));
}

export function decodeHashToPatch(s: string): Patch | null {
  try {
    const json = decompressFromEncodedURIComponent(s);
    if (!json) return null;
    const envelope = JSON.parse(json) as HashEnvelopeV2;
    if (envelope?.v !== PATCH_URL_V || typeof envelope.y !== 'string') return null;
    return patchFromYaml(envelope.y);
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Browser helpers — unified v:2-first, v:1-fallback.
// These replace the v:1-only helpers that used to live in url-state.ts.
// ---------------------------------------------------------------------------

export function readHashState(base: Catalog): WorkingTree | null {
  if (typeof window === 'undefined') return null;
  const hash = window.location.hash;
  if (!hash.startsWith(URL_HASH_PREFIX)) return null;
  const slug = hash.slice(URL_HASH_PREFIX.length);
  return readHashSlug(slug, base);
}

// Exported separately so tests / CLI can feed in a slug without a real
// window.location.hash.
export function readHashSlug(slug: string, base: Catalog): WorkingTree | null {
  const patch = decodeHashToPatch(slug);
  if (patch) return applyPatch(patch, base).tree;
  // Legacy v:1 fallback.
  return decodeTree(slug, base);
}

export function writeHashState(tree: WorkingTree, base: Catalog): void {
  if (typeof window === 'undefined') return;
  const slug = encodePatchToHash(encodePatch(tree, base));
  history.replaceState(null, '', `${URL_HASH_PREFIX}${slug}`);
}

export function buildShareUrl(
  tree: WorkingTree,
  base: Catalog,
): { url: string; length: number; ok: boolean } {
  const slug = encodePatchToHash(encodePatch(tree, base));
  const baseUrl =
    typeof window !== 'undefined' ? window.location.href.split('#')[0] : 'https://example.com/';
  const url = `${baseUrl}${URL_HASH_PREFIX}${slug}`;
  return { url, length: url.length, ok: url.length <= MAX_URL_LENGTH };
}

File: packages/cli-rs/src/patch.rs
use std::collections::HashMap;

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

use crate::diff::{CustomEntry, NodeOverride, PartialRootMeta, TreeDiff, apply_diff};
use crate::tree::path_to_id;
use crate::types::*;

pub const PATCH_SCHEMA: u8 = 1;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatchNode {
    #[serde(rename = "type")]
    pub node_type: NodeType,
    pub name: String,
    pub menu: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub included: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub host: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub probe: Option<Probe>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comparison_children: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extra: Option<IndexMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub children: Option<Vec<PatchNode>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PatchCustomEntry {
    pub parent_path: Option<String>,
    pub node: PatchNode,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Patch {
    pub schema: u8,
    pub base_version: CatalogVersion,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub language: Option<Language>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub root: Option<PartialRootMeta>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub excluded: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub overrides: Option<HashMap<String, NodeOverride>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom: Option<Vec<PatchCustomEntry>>,
}

pub struct BaseMismatch {
    pub patch: CatalogVersion,
    pub actual: Option<CatalogVersion>,
}

pub struct DriftReport {
    pub missing_paths: Vec<String>,
    pub base_mismatch: Option<BaseMismatch>,
}

pub fn apply_patch(patch: &Patch, base: &Catalog) -> (WorkingTree, DriftReport) {
    let mut drift = DriftReport {
        missing_paths: Vec::new(),
        base_mismatch: None,
    };

    if let Some(base_ver) = &base.version {
        if base_ver.sha != patch.base_version.sha {
            drift.base_mismatch = Some(BaseMismatch {
                patch: patch.base_version.clone(),
                actual: Some(base_ver.clone()),
            });
        }
    }

    let mut diff = TreeDiff {
        lang: patch.language.clone(),
        root: patch.root.clone(),
        ex: None,
        ov: None,
        cu: None,
    };

    if let Some(excluded) = &patch.excluded {
        let mut ids = Vec::new();
        for p in excluded {
            match path_to_id(&base.nodes, p) {
                Some(id) => ids.push(id),
                None => drift.missing_paths.push(p.clone()),
            }
        }
        if !ids.is_empty() {
            diff.ex = Some(ids);
        }
    }

    if let Some(overrides) = &patch.overrides {
        let mut ov = HashMap::new();
        for (path, ovr) in overrides {
            match path_to_id(&base.nodes, path) {
                Some(id) => {
                    ov.insert(id, ovr.clone());
                }
                None => drift.missing_paths.push(path.clone()),
            }
        }
        if !ov.is_empty() {
            diff.ov = Some(ov);
        }
    }

    if let Some(custom) = &patch.custom {
        let mut cu = Vec::new();
        for entry in custom {
            let mut parent_id: Option<String> = None;
            if let Some(pp) = &entry.parent_path {
                match path_to_id(&base.nodes, pp) {
                    Some(id) => parent_id = Some(id),
                    None => {
                        drift.missing_paths.push(pp.clone());
                    }
                }
            }
            cu.push(CustomEntry {
                parent_id,
                node: from_patch_node(&entry.node),
            });
        }
        if !cu.is_empty() {
            diff.cu = Some(cu);
        }
    }

    let tree = apply_diff(&diff, base);
    (tree, drift)
}

fn from_patch_node(p: &PatchNode) -> Node {
    Node {
        id: format!("x:{}", uuid::Uuid::new_v4()),
        source: NodeSource::Custom,
        node_type: p.node_type.clone(),
        name: p.name.clone(),
        menu: p.menu.clone(),
        title: p.title.clone(),
        included: p.included.unwrap_or(true),
        children: p
            .children
            .as_ref()
            .map(|c| c.iter().map(from_patch_node).collect())
            .unwrap_or_default(),
        host: p.host.clone(),
        probe: p.probe.clone(),
        comparison_children: p.comparison_children.clone(),
        extra_attrs: p.extra.clone(),
    }
}

pub fn patch_to_yaml(patch: &Patch) -> Result<String, String> {
    serde_yaml_ng::to_string(patch).map_err(|e| e.to_string())
}

pub fn patch_from_yaml(text: &str) -> Result<Patch, String> {
    let parsed: serde_yaml_ng::Value = serde_yaml_ng::from_str(text).map_err(|e| e.to_string())?;

    if !parsed.is_mapping() {
        return Err("not a YAML mapping".to_string());
    }

    let schema = parsed
        .get("schema")
        .and_then(|v| v.as_u64())
        .ok_or_else(|| "missing schema".to_string())?;

    if schema != PATCH_SCHEMA as u64 {
        return Err(format!(
            "unsupported schema (got {}, expected {})",
            schema, PATCH_SCHEMA
        ));
    }

    let base_version = parsed.get("baseVersion");
    if base_version.is_none()
        || !base_version
            .unwrap()
            .get("sha")
            .map(|v| v.is_string())
            .unwrap_or(false)
    {
        return Err("missing or malformed baseVersion".to_string());
    }

    serde_yaml_ng::from_str(text).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tree::find_node;

    static CATALOG_JSON: &str = include_str!(concat!(env!("OUT_DIR"), "/catalog.json"));

    fn catalog() -> Catalog {
        serde_json::from_str(CATALOG_JSON).unwrap()
    }

    fn version() -> CatalogVersion {
        catalog().version.unwrap()
    }

    fn minimal_patch() -> Patch {
        Patch {
            schema: PATCH_SCHEMA,
            base_version: version(),
            language: None,
            root: None,
            excluded: None,
            overrides: None,
            custom: None,
        }
    }

    #[test]
    fn fresh_tree_produces_minimal_patch_yaml() {
        let patch = minimal_patch();
        let yaml = patch_to_yaml(&patch).unwrap();
        assert!(yaml.contains("schema: 1"));
        assert!(yaml.contains("baseVersion"));
    }

    #[test]
    fn apply_minimal_patch_no_drift() {
        let cat = catalog();
        let patch = minimal_patch();
        let (tree, drift) = apply_patch(&patch, &cat);
        assert!(drift.missing_paths.is_empty());
        assert!(drift.base_mismatch.is_none());
        assert_eq!(tree.nodes.len(), cat.nodes.len());
    }

    #[test]
    fn apply_patch_with_exclusion() {
        let cat = catalog();
        let patch = Patch {
            excluded: Some(vec!["/CDN/Cloudflare".to_string()]),
            ..minimal_patch()
        };
        let (tree, drift) = apply_patch(&patch, &cat);
        assert!(drift.missing_paths.is_empty());
        let cf = find_node(&tree.nodes, "c:CDN/Cloudflare").unwrap();
        assert!(!cf.included);
    }

    #[test]
    fn apply_patch_with_override() {
        let cat = catalog();
        let mut overrides = HashMap::new();
        overrides.insert(
            "/CDN/Cloudflare".to_string(),
            NodeOverride {
                host: Some(Some("1.1.1.1".to_string())),
                ..Default::default()
            },
        );
        let patch = Patch {
            overrides: Some(overrides),
            ..minimal_patch()
        };
        let (tree, _) = apply_patch(&patch, &cat);
        let cf = find_node(&tree.nodes, "c:CDN/Cloudflare").unwrap();
        assert_eq!(cf.host.as_deref(), Some("1.1.1.1"));
    }

    #[test]
    fn missing_paths_reported_as_drift() {
        let cat = catalog();
        let patch = Patch {
            excluded: Some(vec!["/CDN/DoesNotExist".to_string()]),
            ..minimal_patch()
        };
        let (_, drift) = apply_patch(&patch, &cat);
        assert_eq!(drift.missing_paths, vec!["/CDN/DoesNotExist"]);
    }

    #[test]
    fn base_version_mismatch_reported() {
        let cat = catalog();
        let patch = Patch {
            base_version: CatalogVersion {
                date: "2020-01-01".to_string(),
                sha: "aaaaaaa".to_string(),
            },
            ..minimal_patch()
        };
        let (_, drift) = apply_patch(&patch, &cat);
        let mm = drift.base_mismatch.unwrap();
        assert_eq!(mm.patch.sha, "aaaaaaa");
    }

    #[test]
    fn no_base_mismatch_when_sha_matches() {
        let cat = catalog();
        let patch = minimal_patch();
        let (_, drift) = apply_patch(&patch, &cat);
        assert!(drift.base_mismatch.is_none());
    }

    #[test]
    fn yaml_round_trip() {
        let patch = Patch {
            excluded: Some(vec!["/CDN/Akamai".to_string()]),
            ..minimal_patch()
        };
        let yaml = patch_to_yaml(&patch).unwrap();
        let parsed = patch_from_yaml(&yaml).unwrap();
        assert_eq!(parsed.schema, PATCH_SCHEMA);
        assert_eq!(parsed.excluded.as_ref().unwrap(), &["/CDN/Akamai"]);
    }

    #[test]
    fn patch_from_yaml_rejects_wrong_schema() {
        let yaml = "schema: 99\nbaseVersion:\n  date: x\n  sha: y\n";
        let err = patch_from_yaml(yaml).unwrap_err();
        assert!(err.contains("unsupported schema"));
    }

    #[test]
    fn patch_from_yaml_rejects_missing_base_version() {
        let yaml = "schema: 1\n";
        let err = patch_from_yaml(yaml).unwrap_err();
        assert!(err.contains("baseVersion"));
    }

    #[test]
    fn custom_nodes_round_trip_through_yaml() {
        let patch = Patch {
            custom: Some(vec![PatchCustomEntry {
                parent_path: None,
                node: PatchNode {
                    node_type: NodeType::Category,
                    name: "MyStuff".to_string(),
                    menu: "My Stuff".to_string(),
                    title: "Personal".to_string(),
                    included: None,
                    host: None,
                    probe: None,
                    comparison_children: None,
                    extra: None,
                    children: None,
                },
            }]),
            ..minimal_patch()
        };
        let yaml = patch_to_yaml(&patch).unwrap();
        let parsed = patch_from_yaml(&yaml).unwrap();
        let custom = parsed.custom.unwrap();
        assert_eq!(custom.len(), 1);
        assert_eq!(custom[0].node.name, "MyStuff");
    }
}
